### ISSUE TITLE
Transaction exceeding ttl does not release the lock

### ISSUE THREAD

**Issue by lxr599** [pingcap/tidb#49184]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

 ```sql
  t1                    |                    t2
  
begin;    

select * from t1;
+------+------+
| a    | b    |
+------+------+
|    3 |    2 |
|    2 |    3 |
+------+------+
                                        begin;    

                                        select * from t1;
                                        +------+------+
                                        | a    | b    |
                                        +------+------+
                                        |    3 |    2 |
                                        |    2 |    3 |
                                        +------+------+

                                        delete from t1 where a=3;                                     
                                        

update t1 set a=4 where a=3;

                                        select * from t1;                                     
										
                                        commit;                                       

select * from t1;

commit;
 ```

The codes are modified from @zyguan's code , and here is the [link](https://gist.github.com/zyguan/bf2ba336fc7902f3b279535417905554).

```javascript
Deno.test('test', async () => {
  const c1 = await mysql.createConnection({ host: TIDB_HOST, port: 4000, user: "root", database: 'lxrdb' });
  const c2 = await mysql.createConnection({ host: TIDB_HOST, port: 4000, user: "root", database: 'lxrdb' });
  try {
    // await c1.query("set @@tidb_general_log=1, @@innodb_lock_wait_timeout=600");
    await c1.query("set @@tidb_general_log=1");
    await c1.query("drop table if exists t1");
    await c1.query("create table t1 (a int, b int)");
    await c1.query("insert into t1 values (3, 2), (2, 3)");
    log("show config", await pp(c1.query("show config where name like '%max-txn-ttl'")));

    await c1.query("begin");
    log("t1 select(o1):", await pp(c2.query("select * from t1")));
      
    await c2.query("begin");
    log("t2 select:", await pp(c2.query("select * from t1")));
    log("t2 delete:", await pp(c2.query("delete from t1 where a=3")));

    log("t1 update(o2):", await pp(c1.query("update t1 set a=4 where a=3")));
      
    log("t2 select:", await pp(c2.query("select * from t1")));
      
    log("t1 select(o3):", await pp(c1.query("select * from t1")));
    log("t1 select(o3'):", await pp(c1.query("select * from t1")));
      
    log("t2 commit:", await pp(c2.query("commit")));
      
    log("t1 commit:", await pp(c1.query("commit")));

  } finally {
    await close(c1);
    await close(c2);
  }
});
```

### 2. What did you expect to see? (Required)

Since we set `max-txn-ttl` to 30s, and `innodb_lock_wait_timeout` is 50s. The `update` in t1 will wait for the `delete` in t2 to release the lock. Theoretically, `update` will be executed successfully after waiting for 30 seconds since t2 exceeds the ttl. 

### 3. What did you see instead (Required)

The `update` in t1 will return two result:

1. `update` was executed successfully, but took about 35 seconds.
2. `update` returns an error: "MySQLTransactionRollbackException: Lock wait timeout exceeded; try restarting transaction".

It seems that result 2 is unexpected. Since this error is returned, it means that the waiting time of t1 exceeds 50 seconds, and this time has exceeded ttl, which is 30 seconds, so t2 should theoretically have released the lock.

### 4. What is your TiDB version? (Required)

tidb_version(): Release Version: v7.4.0
Edition: Community
Git Commit Hash: 38cb4f3312be9199a983c0ef282d2ea2e28a7824
Git Branch: heads/refs/tags/v7.4.0
UTC Build Time: 2023-10-10 14:18:50
GoVersion: go1.21.1
Race Enabled: false
Check Table Before Drop: false
Store: tikv

**Comment by lxr599** [comment]
This bug may be considered to be the jdbc version of #48907.

**Comment by zyguan** [comment]
Thank you for your report, @lxr599.
Although the phenomenon is weird, it's actually by design in tidb. The TTL keepalive routine works as the following:

```
for {
  wait 10s
  let uptime = now() - start-ts
  if uptime > max-txn-ttl {
    mark txn timeout
    break
  }
  push ttl to (uptime + 20s)
}
```

So, in this case, there might be two outcomes:
1. t2's keepalive loop iterates 3 times, at last time, uptime is slightly greater than max-txn-ttl (eg. 30050), then t1 can resolve the lock after ~40s
![2023-12-05_165647](https://github.com/pingcap/tidb/assets/6850317/de0a694c-6554-40e5-90de-9b6d78e63a03)
2. t2's keepalive loop iterates 4 times, at 3rd time, uptime is equal to max-txn-ttl, then t2's ttl will be pushed to 50s and one more iteration is needed, so t1 cannot resolve it in default innodb_lock_wait_timeout
![2023-12-05_170043](https://github.com/pingcap/tidb/assets/6850317/c7e841da-40fe-4ace-97e6-3c3fcc116314)

**Comment by wengsy150943** [comment]
> Thank you for your report, @lxr599. Although the phenomenon is weird, it's actually by design in tidb. The TTL keepalive routine works as the following:
> 
> ```
> for {
>   wait 10s
>   let uptime = now() - start-ts
>   if uptime > max-txn-ttl {
>     mark txn timeout
>     break
>   }
>   push ttl to (uptime + 20s)
> }
> ```
> 
> So, in this case, there might be two outcomes:
> 
> 1. t2's keepalive loop iterates 3 times, at last time, uptime is slightly greater than max-txn-ttl (eg. 30050), then t1 can resolve the lock after ~40s
> 2. t2's keepalive loop iterates 4 times, at 3rd time, uptime is equal to max-txn-ttl, then t2's ttl will be pushed to 50s and one more iteration is needed, so t1 cannot resolve it in default innodb_lock_wait_timeout

Thanks for your reply.
However, the 2nd type behaviour seems still weird according to your description.
When t2's keepalive loop iterates at its 3rd time, the ttl is pushed to 50s. But the next iterate should be ran at 40s, since it only waits for 10s(The screen shot also shows that a log is printed when ttl=40s).
![image](https://github.com/pingcap/tidb/assets/43516416/bcbf1b66-1b7a-4f02-b8af-43ddb6c32510)

Based on your code, there will be a check (` if uptime > max-txn-ttl `) in each iterate. So t2's keepalive loop should check if it  exceeds max-txn-ttl at 40s. And it does exceed so that t2 should be timeout at that time(40s). However, 40s is **less** that lock_wait_timeout so t1 **can** resolve the lock. But in case 2, t1 **cannot** resolve the lock.

**Comment by zyguan** [comment]
Sorry for not making it clear. I guess there might be some misunderstandings about txn ttl and lock ttl
- keepalive loop here does 2 things for each iteration
  1. check the txn ttl, break on uptime > max-txn-ttl
  2. push the lock ttl to uptime+20s if uptime <= max-txn-ttl 
- when uptime > max-txn-ttl, the txn will be marked as timeout
  - the txn is still alive (neither be committed nor be rollbacked) even when it has been marked as timeout
  - you need to commit/rollback the txn manually, tidb won't do it for you (as described [here](https://github.com/pingcap/tidb/issues/48907#issuecomment-1835555558))
  - other txns with write conflicts can resolve locks only after these locks expire (instead of the txn timeout)

> Based on your code, there will be a check (if uptime > max-txn-ttl) in each iterate. So t2's keepalive loop should check if it exceeds max-txn-ttl at 40s. And it does exceed so that t2 should be timeout at that time(40s). However, 40s is less that lock_wait_timeout so t1 can resolve the lock. But in case 2, t1 cannot resolve the lock.

Here t2 timed out at uptime=40s, but it's still alive and its locks would be valid until 50s, so t1 cannot resolve the lock.

If you let t2 be committed/rollbacked at 45s, like the following, you will find that t1's update will always be executed successfully:
1. at ~40s (1st case), lock timed out at 40s and was resolved by t1
   ![2023-12-19_121809](https://github.com/pingcap/tidb/assets/6850317/cc876882-7e4a-41e4-93db-eb31170e2001)
2. at ~45s (2nd case), lock was released by t2 earlier (otherwise timed out at 50s)
   ![2023-12-19_121840](https://github.com/pingcap/tidb/assets/6850317/557f2cd0-4962-4f7b-8eb7-da311a599988)

```js
    await c1.query("drop table if exists t1");
    await c1.query("create table t1 (a int, b int)");
    await c1.query("insert into t1 values (3, 2), (2, 3)");
    log("show max-txn-ttl", await pp(c1.query("show config where name like '%max-txn-ttl'")));
    log("show lock_wait_timeout", await pp(c1.query("select @@innodb_lock_wait_timeout")));

    await c1.query("begin");
    log("t1 select(o1):", await pp(c1.query("select * from t1")));

    await c2.query("begin");
    log("t2 select:", await pp(c2.query("select * from t1")));
    log("t2 delete:", await pp(c2.query("delete from t1 where a=3")));

    const t2 = new Promise((resolve) => {
      setTimeout(async () => {
        log("t2 select:", await pp(c2.query("select * from t1")));
        log("t2 commit:", await pp(c2.query("commit")));
        resolve();
      }, 45000);
    });

    log("t1 update(o2):", await pp(c1.query("update t1 set a=4 where a=3")));
    log("t1 select(o3):", await pp(c1.query("select * from t1")));
    log("t1 select(o3'):", await pp(c1.query("select * from t1")));
    log("t1 commit:", await pp(c1.query("commit")));

    await t2;
```
